### ISSUE TITLE
Reduce number of running containers in CI to one per run

### ISSUE THREAD

**Issue by jkaflik** [ClickHouse/clickhouse-go#1406]
## Observed

Each CI tests run creates a few ClickHouse containers per test directory. This is an unnecessary overhead that could be reduced to a single container per run.

## Expected behaviour

Run a single ClickHouse container per tests run.

**Comment by romanchechyotkin** [comment]
hello, have you thought about running docker container in CI and make connections to this clickhouse container?

**Comment by jkaflik** [comment]
CI is using running ClickHouse in container.

**Comment by romanchechyotkin** [comment]
i meant use `services` https://docs.github.com/en/actions/use-cases-and-examples/using-containerized-services/about-service-containers
in tests we just connect to container running during CI, also we can mark tests with tag, which are using clickhouse for testing

**Comment by jkaflik** [comment]
@romanchechyotkin no, as it's going for specific CI runners. We expect the test to be run with `go test` without any other requirements. Thus, test spawns container itself. 

We don't want to maintain two ways of running dependencies.

**Comment by tosolveit** [comment]
Hi everyone,

I've been exploring ways to reduce the container startup overhead during test runs and wanted to share a summary of what I've tried or researched — along with a suggestion and a couple of open questions.

---

### 🚧 Problem

Currently, each test package spins up its own `clickhouse/clickhouse-server` container via `testcontainers-go`, which introduces significant overhead — especially in CI, where test packages run in separate binaries.

---

### 🔍 What I’ve explored

#### 1. **Shared `Runtime()` across packages**

✅ **Tried this**

I pointed all `TestMain()` functions to the same `Runtime()` logic with `sync.Once` to create the container only once.
Unfortunately, this doesn’t work: each test package is compiled and run as a separate binary by Go's test runner, so globals like `sync.Once` don’t coordinate across packages.

#### 2. **File lock or shared env var (e.g., `CLICKHOUSE_TEST_CONTAINER_ID`)**

🧐 **Researched**

* Tried locking with `os.OpenFile(..., syscall.LOCK_EX)` or sharing a container ID through a file/env var.
* ❌ Brittle and error-prone: requires careful file permissions, cleanup, and is hard to debug if concurrent access causes race conditions.

#### 3. **Use a persistent container via `docker-compose`**

🧐 **Researched**

Since `docker-compose.yml` already exists in the repo, we could do:

```bash
docker-compose up -d clickhouse
CLICKHOUSE_HOST=localhost go test ./...
```

This would share one container across all packages, reducing startup time.
But it comes with trade-offs:

* ❌ No test isolation (tests could interfere via shared state)
* ❌ No auto-cleanup (requires `docker-compose down`)
* ❌ Fixed ports, no dynamic mapping

#### 4. **Refactor all tests into a single package**

🧐 **Researched**

We could ensure a single `TestMain()` manages the container by placing all tests in one package.
However:

* ❌ Requires large-scale refactoring
* ❌ Breaks modularity and test ownership
* ❌ Harder to navigate and extend

#### 5. **Other ways of achieving this**
 I don't know.

---

### ✅ Current Thinking

Although there's duplication, I now believe the current `testcontainers-go` setup offers the best trade-off:

* ✅ Isolated containers per test suite
* ✅ Reliable in both CI and local
* ✅ Easy to reason about and extend

---

### 🙏 Questions

* Would maintainers be open to a faster `docker-compose` mode (e.g., gated behind `CLICKHOUSE_USE_COMPOSE=true`) for local development?
* Do you think full refactoring into one test package is worth considering? Or is modularity too valuable to sacrifice?
* Or — do you also think our current setup, while duplicative, provides the best overall developer experience?

Thanks a lot for your time and input! Happy to iterate on any of these approaches. 🙏

---

Let me know if you'd like me to open a PR or test out any specific direction.
